File: 2015/Day 1 Not Quite Lisp/main.py
# pylint: disable-all
# Import AOC Common
import os
import sys
import time
sys.path.append("../")
from shared import aoc_common #, aoc_algorithms, aoc_grid_tools


P1_DEBUG = True
P2_DEBUG = True

USE_REAL_DATA = True # Loads input.txt when True or sample.txt when False

if USE_REAL_DATA: aoc_common.get_aoc_puzzle_data()
INPUT_FILENAME  = "%s/input.txt" % os.path.dirname(os.path.realpath(__file__))
SAMPLE_FILENAME = "%s/sample.txt" % os.path.dirname(os.path.realpath(__file__))


def part1(input):
    target_floor = 0 # ground floor
    for char in [*input[0]]:
        match char:
            case "(":
                if P1_DEBUG: print("Going UP 1 floor")
                target_floor += 1
            case ")":
                if P1_DEBUG: print("Going DOWN 1 floor")
                target_floor -= 1

    return target_floor


def part2(input):
    target_floor = 0 # ground floor
    for index, char in enumerate([*input[0]]):
        match char:
            case "(":
                if P1_DEBUG: print("Going UP 1 floor")
                target_floor += 1
            case ")":
                if P1_DEBUG: print("Going DOWN 1 floor")
                target_floor -= 1

        if target_floor == -1:
            if P1_DEBUG: print("Entering Basement...")
            return index+1


if __name__ == '__main__':
    parsed_input = aoc_common.import_file_single_new_line(INPUT_FILENAME if USE_REAL_DATA else SAMPLE_FILENAME)

    start_time_part1 = time.time()
    part_1 = part1(parsed_input)
    end_time_part1 = time.time()

    start_time_part2 = time.time()
    part_2 = part2(parsed_input)
    end_time_part2 = time.time()

    print("# # # SOLUTIONS # # #")
    print(f"Part 1: {part_1} \t ⏱️ in {end_time_part1 - start_time_part1:.4f} seconds")
    print(f"Part 2: {part_2} \t ⏱️ in {end_time_part2 - start_time_part2:.4f} seconds")

File: 2015/Day 1 Not Quite Lisp/main.go
package main

import (
	"aocshared"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

const (
	P1_DEBUG      = true
	P2_DEBUG      = true
	USE_REAL_DATA = true // Use input.txt when true, sample.txt when false
)

func part1(input []string) string {
	targetFloor := 0

	for _, char := range input[0] {
		switch char {
		case '(':
			if P1_DEBUG {
				fmt.Println("Going UP 1 floor")
			}
			targetFloor++
		case ')':
			if P1_DEBUG {
				fmt.Println("Going DOWN 1 floor")
			}
			targetFloor--
		}
	}

	return fmt.Sprintf("%d", targetFloor)
}

func part2(input []string) string {
	targetFloor := 0

	for i, char := range input[0] {
		switch char {
		case '(':
			if P2_DEBUG {
				fmt.Println("Going UP 1 floor")
			}
			targetFloor++
		case ')':
			if P2_DEBUG {
				fmt.Println("Going DOWN 1 floor")
			}
			targetFloor--
		}

		if targetFloor == -1 {
			if P2_DEBUG {
				fmt.Println("Entering Basement...")
			}
			return fmt.Sprintf("%d", i+1)
		}
	}

	return "-1"
}

func main() {
	// Determine which file to use
	filename := "sample.txt"
	if USE_REAL_DATA {
		filename = "input.txt"
	}

	// Resolve input from the current directory first (works with go run),
	// then fall back to the executable directory.
	workingDir, err := os.Getwd()
	if err != nil {
		panic(err)
	}
	filePath := filepath.Join(workingDir, filename)
	if _, err := os.Stat(filePath); err != nil {
		ex, err := os.Executable()
		if err != nil {
			panic(err)
		}
		exePath := filepath.Dir(ex)
		filePath = filepath.Join(exePath, filename)
	}

	// Parse input
	parsedInput, err := aocshared.ImportFileSingleNewLine(filePath)
	if err != nil {
		panic(err)
	}

	// Part 1
	startTimePart1 := time.Now()
	part1Result := part1(parsedInput)
	endTimePart1 := time.Now()

	// Part 2
	startTimePart2 := time.Now()
	part2Result := part2(parsedInput)
	endTimePart2 := time.Now()

	// Output results
	fmt.Println("# # # SOLUTIONS # # #")
	fmt.Printf("Part 1: %s \t ⏱️  in %.4f seconds\n", part1Result, endTimePart1.Sub(startTimePart1).Seconds())
	fmt.Printf("Part 2: %s \t ⏱️  in %.4f seconds\n", part2Result, endTimePart2.Sub(startTimePart2).Seconds())
}
